rfc/005-soft-chain-upgrades.md

```diff
@@ -0,0 +1,227 @@
+
+# RFC 005: Soft Chain Upgrades
+
+## Changelog
+
+- 2020-11-16: Initial draft.
+
+## Author(s)
+
+- Callum Waters (@cmwaters)
+
+## Context
+
+The current method for upgrading a network across block protocol breaking
+changes is to hard fork the chain. All nodes must coordinate a halt height/
+time, where application state can be exported across and an entirely new chain
+(as far as Tendermint is concerned) is created (see [cosmoshub-3
+upgrade](https://blog.cosmos.network/cosmos-hub-3-upgrade-announcement-39c9da941aee)).
+This means that each block protocol version maps to the entire life of a single
+chain. The result of this is that either Tendermint is hindered with the
+flexibility and quantity of changes it can make or that networks have to deal
+with pain of frequently upgrading.
+
+Furthermore, although work from [RFC002: Non-Zero Genesis](https://github.com/tendermint/spec/blob/master/rfc/002-nonzero-genesis.md)
+has enabled unique heights throughout the history of a "logical" chain, data
+retrievability across heights remains difficult as blockchains that choose to
+upgrade not only have to coordinate upgrades but they essentially lose all their
+history.
+
+The scope of this RFC is to address the need for Tendermint to have flexibility
+in improving itself whilst offering greater ease of use to the networks running
+on Tendermint. This is done by introducing a division in the types of ways a
+network can be upgraded. There will still remain the need for hard upgrades (the
```

Comment:
> Why will there still remain a need for hard upgrades?

Because I can't bind what future tendermint developers do nor predict how tendermint will evolve.

What I've attempted to do is to split the problem by identifying a set of data structures that are likely to change and give them flexibility to change without requiring a "hard break" or a reset of the block chain. At the moment, Tendermint releases a major version every 6 months but this might allow major versions to only happen every couple of years, thus extending the lifetime of a blockchain. But this is still an experimental field, we could decide to completely change up the consensus mechanism and in that case v2 could look completely different from v1. I find it unreasonable to have a single software version that tries to tie all this together. 

You're right though that we should also be addressing when a major change happens how we retain the appearance of a full immutable history that can be accessed and I've been trying to come up with ideas here https://github.com/tendermint/tendermint/issues/5595#issuecomment-724807340 to essentially improve the ux of a "hard upgrade"

Reply:
> I find it unreasonable to have a single software version that tries to tie all this together.

I sort of disagree with this. It depends on what kind of product we want to offer out users, but people don't have to throw their data away when upgrading e.g. Postgres 8 to 13, and I think we should be able to do the same. It doesn't have to be convenient (e.g. a migration process is totally fine), but I think we should offer some way for people to keep their data across upgrades.

Reply:
yeah I did like the idea of turning nodes that were used in older versions into archive nodes which only run say the blockchain reactor and an rpc

Reply:
basically during a major version change, the old node would pass the "baton" i.e. config + genesis to the new node to start up, but it wouldn't quite be left to throw away. Instead it would become an archive node serving data from that "era"

Reply:
Yeah, that might work. Problem is that this requires coordination with all users, because they now have to know which heights are located at which servers, and at what height a future upgrade is taking place. If this is the approach we're planning to use, then I feel like we should have some sort of built-in frontend proxy that automatically does this mapping, such that when a network upgrades the governors can make sure all of these proxies have been updated with new network information and relevant block ranges before the upgrade takes place. This way, users/clients don't have to do anything.

Of course, this doesn't solve the problem of having to support old RPC versions for new blocks, since this will still require clients to upgrade or suddenly break. Maybe that's okay once in a while, it depends on the guarantees we want to offer users. But worth thinking through how we want this to work.

Reply:
Yup, I agree that ideally we want nothing to change from the perspective of the external client.

The quasi "brute-force" solution I see is that all nodes import their previous versions. That way the latest version is something of a "master" node and, knowing the upgrade heights, can redirect requests to the corresponding archive node. You could probably get away with just running an extra blockchain reactor and have the same rpc wired to that. I'm just a bit concerned that someone that starts a network at v2.3 has this unnecessarily big binary that he/she uses.